### PR TITLE
Incremental improvements for non-released ModelsRepository client.

### PR DESCRIPTION
**Pull request by digimaun**
* Changes simplest ctor to use dependency resolution option of
  TryFromExpanded. We can guarantee the global endpoint has expanded
  form for any model.
* Hides the existing ModelsRepositoryClient.IsValidDtmi(). We will
  revisit exposing after more real-world usage.
* If input dtmis to Resolve() are invalid thrown exception changed from `RequestFailedException` to `ArgumentException`.
* Add some polish/naming updates and docstring improvements.
* Adds samples content including integration with Parser.
* Have more TODOs in the samples arena including the README.
* Have more TODOs in Client surface area with respect to API review feedback. Like rename of Resolve().

### CODE REVIEW

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            // The client will also work with a local filesystem Uri. This example shows initalization
+            // with a local Uri and disabling model dependency resolution.
+            var localRepoPath = ClientSamplesLocalModelsRepository;
+            var localRepoClient = new ModelsRepositoryClient(
+                new Uri(localRepoPath),
+                new ModelsRepositoryClientOptions(resolutionOption: DependencyResolutionOption.Disabled));
+            Console.WriteLine($"Initialized client pointing to local path: {localRepoClient.RepositoryUri}");
+        }
+
+        public static async Task ResolveExistingModelsFromEndpointAsync()
```
Comment (azabbasi) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs:40]: Now that we are changing the API name to GetModels, I would rename these methods too.
also `FromEndpoint` doesn't tell me much, maybe `FromGlobalRepo` or `FromDefaultRepo` ?

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            var dtmi = "dtmi:com:example:TemperatureController;1";
+
+            // Global endpoint client
+            var client = new ModelsRepositoryClient();
+
+            // The output of ResolveAsync() will include at least the definition for the target dtmi.
```
Comment (azabbasi) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs:47]: Method name change should be applied to this as well.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs [hunk 3]
```diff
@@ -4,22 +4,23 @@
...
+        /// Enable model dependency resolution.
         /// </summary>
         Enabled,
 
         /// <summary>
-        /// Try to get external dependencies using .expanded.json.
+        /// Try to get pre-computed model dependencies using .expanded.json.
+        /// If the model expanded form does not exist fall back to normal dependency processing.
         /// </summary>
         TryFromExpanded
```
Comment (azabbasi) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:25]: if this is the default option (which I think it is) let's move it to the top of the enum declaration.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs [hunk 4]
```diff
@@ -21,14 +21,15 @@ public class ModelsRepositoryClient
         private readonly ModelsRepositoryClientOptions _clientOptions;
 
         /// <summary>
-        /// Initializes the ModelsRepositoryClient with default client options while pointing to
-        /// the Azure IoT Plug and Play Models repository https://devicemodels.azure.com for resolution.
+        /// Initializes the ModelsRepositoryClient to point to the Azure IoT Models Repository https://devicemodels.azure.com
```
Comment (azabbasi) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs:24]: <see href="https://devicemodels.azure.com">Azure IoT Models Repository</see>

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs [hunk 5]
```diff
@@ -21,14 +21,15 @@ public class ModelsRepositoryClient
...
+        /// with the model dependency resolution option of TryFromExpanded.
         /// </summary>
-        public ModelsRepositoryClient() : this(DefaultModelsRepository, new ModelsRepositoryClientOptions()) { }
+        public ModelsRepositoryClient() : this(DefaultModelsRepository,
+            new ModelsRepositoryClientOptions(resolutionOption: DependencyResolutionOption.TryFromExpanded)) { }
 
         /// <summary>
         /// Initializes the ModelsRepositoryClient with custom client <paramref name="options"/> while pointing to
-        /// the Azure IoT Plug and Play Model repository https://devicemodels.azure.com for resolution.
+        /// the Azure IoT Models Repository https://devicemodels.azure.com.
```
Comment (azabbasi) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs:32]: same here 
`<see href="https://devicemodels.azure.com">Azure IoT Models Repository</see>`

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClientOptions.cs [hunk 6]
```diff
@@ -14,7 +14,7 @@ public class ModelsRepositoryClientOptions : ClientOptions
         internal const ServiceVersion LatestVersion = ServiceVersion.V2021_02_11;
 
         /// <summary>
-        /// The versions of Azure IoT Model Repository by this client
+        /// The versions of Azure IoT Models Repository supported by this client.
```
Comment (azabbasi) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClientOptions.cs:17]: add a service in there :D 
`/// The versions of Azure IoT Models Repository service supported by this client.`

**Review by azabbasi — APPROVED** [review]

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs (hunk 2 above)
Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs:47]: same everywhere.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    internal class ModelResolutionSamples
+    {
+        private static string ClientSamplesDirectoryPath => Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+        private static string ClientSamplesLocalModelsRepository => Path.Combine(ClientSamplesDirectoryPath, "SampleModelsRepo");
+
+        public static void ClientInitialization()
+        {
+            // When no Uri is provided for instantiation. The global Azure IoT Models Repository endpoint is used
+            // and the dependency model resolution option is set to TryFromExpanded.
+            var globalRepoClient = new ModelsRepositoryClient();
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs:22]: The advice is just to name it `client`.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            Console.WriteLine($"Initialized client pointing to global endpoint: {globalRepoClient.RepositoryUri}");
+
+            // This form shows specifing a custom Uri for the models repository with default client options.
+            // The default client options will enable model dependency resolution.
+            var remoteRepoEndpoint = "https://contoso.com/models";
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs:27]: const

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs (hunk 1 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs]: Made the `FromEndpoint` -> `FromGlobalRepo` change as you suggested. But if you don't mind, will take care of the `Resolve()` -> `GetModels()` change in next PR.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs (hunk 3 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs]: Done.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs (hunk 4 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs]: Good idea done.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs (hunk 5 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs]: Done.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClientOptions.cs (hunk 6 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClientOptions.cs]: Good call!!

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs (hunk 8 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs]: done

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs (hunk 7 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs]: done.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs (hunk 2 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/ModelResolutionSamples.cs:45]: Got it - will apply in upcoming method name change PR.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs [hunk 9]
```diff
@@ -20,34 +18,19 @@ public static async Task Main(string[] args)
...
-        private static async Task TryResolveButNotFoundAsync()
-        {
-            var dtmi = "dtmi:com:example:NotFound;1";
-            var client = new ModelsRepositoryClient();
+            // Model Resolution samples
+            await ModelResolutionSamples.ResolveExistingModelsFromGlobalRepoAsync();
+            await ModelResolutionSamples.ResolveExistingModelsFromLocalRepoAsync();
+            await ModelResolutionSamples.TryResolveFromGlobalRepoButNotFoundAsync();
+            await ModelResolutionSamples.TryResolveFromLocalRepoButNotFoundAsync();
+            await ModelResolutionSamples.TryResolveWithInvalidDtmi();
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs:29]: Needs `Async` suffix

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs [hunk 10]
```diff
@@ -4,23 +4,24 @@
...
     /// </summary>
     public enum DependencyResolutionOption
     {
         /// <summary>
-        /// Do not process external dependencies.
+        /// Try to get pre-computed model dependencies using .expanded.json.
+        /// If the model expanded form does not exist fall back to normal dependency processing.
         /// </summary>
-        Disabled,
+        TryFromExpanded,
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:15]: I preferred the previous order, in that it seemed to grow in functionality, and Disabled (0) was the default element.

Why the change?

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs [hunk 11]
```diff
@@ -4,23 +4,24 @@
...
     /// <summary>
-    /// The dependency processing options.
+    /// The model dependency resolution options.
     /// </summary>
     public enum DependencyResolutionOption
     {
         /// <summary>
-        /// Do not process external dependencies.
+        /// Try to get pre-computed model dependencies using .expanded.json.
+        /// If the model expanded form does not exist fall back to normal dependency processing.
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:13]: What is normal processing that we fall back to?

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/FetchResult.cs [hunk 12]
```diff
@@ -15,7 +15,7 @@ internal class FetchResult
         public string Path { get; set; }
 
         public bool FromExpanded => Path.EndsWith(
-            ModelRepositoryConstants.ExpandedJsonFileExtension,
+            ModelsRepositoryConstants.ExpandedJsonFileExtension,
             System.StringComparison.InvariantCultureIgnoreCase);
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/FetchResult.cs:19]: Prefer move `System` to `using System;` above.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs (hunk 10 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:15]: The change was from this feedback: https://github.com/Azure/azure-sdk-for-net/pull/19042#discussion_r583235811
And there are more implications like explicitly setting the enum integers since the processing logic depends on it.

I'll put it back in my upcoming PR.

Comment (drwill-ms) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:15]: @azabbasi has a good point, but I think the new order is odd. I don't feel very strong about this, so I'll stay with just noting to you both that the order seems weird to me. If you both disagree, we can keep it as is. If you prefer to change it after reconsidering, that is fine too.

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:15]: I didn't know that Disabled is the default element. in the comment I mentioned that if TryFromExpanded is the default behavior, we should have it at the top.

Comment (drwill-ms) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:15]: I'm not saying Disabled is the default element; it just made sense to me that Disabled should be the default element in an enum.

It sounds like the default for the ctor (if not specified) is expanded though, which is what you originally keyed off of.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs (hunk 11 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:13]: Updated the wording a bit for this enum which tries to explain this better in the recently opened PR https://github.com/Azure/azure-sdk-for-net/pull/19072

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/FetchResult.cs (hunk 12 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/FetchResult.cs:19]: Done in recently opened PR https://github.com/Azure/azure-sdk-for-net/pull/19072

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs (hunk 10 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/DependencyResolutionOption.cs:15]: Let's move discussion to recently opened PR https://github.com/Azure/azure-sdk-for-net/pull/19072

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs (hunk 9 above)
Comment (digimaun) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs:29]: Done https://github.com/Azure/azure-sdk-for-net/pull/19072
